### ISSUE TITLE
Break compliance ties by draft recency in the UI

### ISSUE THREAD

**Issue by Julian** [bowtie-json-schema/bowtie#677]
A page like https://bowtie.report/#/implementations/java-json-schema which lists all supported dialects for the implementation displays a table of compliance with those dialects.

It is currently sorted by compliance (i.e. fewer failed tests at the top) -- but it should break ties by publication date of the draft, such that the top row is the newest draft.

**Comment by AK4266** [comment]
Hey @Julian  I would like to work on this issue. Can you assign it to me

**Comment by Julian** [comment]
Great! Done. Ask if you have any questions!

**Comment by Julian** [comment]
@AK4266 also feel free to come [join the JSON Schema Slack](https://json-schema.org/slack) if you haven't already!

**Comment by sanskar-soni-9** [comment]
Hi @Julian, I would like to work on this issue.

I looked at the issue and I don't think its currently sorted cuz every time I refresh it renders it in random order and not in any sorted order.

**Comment by Julian** [comment]
Great, I've assigned you!

> I looked at the issue and I don't think its currently sorted cuz every time I refresh it renders it in random order and not in any sorted order.

You look correct! It's only on the dialect report page that we sort at all (see [here](
https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/components/Summary/SummaryTable.tsx#L61-L78), on the implementation page we don't (see [here](
https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/components/ImplementationReportView/DialectCompliance.tsx#L30-L40)

This ticket is about the latter, so we should do the sort first by compliance (same as in the first link) and then break ties by what is in the description.

Ask if you run into any other trouble!

**Comment by sanskar-soni-9** [comment]
> Great, I've assigned you!
> 
> > I looked at the issue and I don't think its currently sorted cuz every time I refresh it renders it in random order and not in any sorted order.
> 
> You look correct! It's only on the dialect report page that we sort at all (see [here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/components/Summary/SummaryTable.tsx#L61-L78)) but on the implementation page we indeed don't (see [here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/components/ImplementationReportView/DialectCompliance.tsx#L30-L40)).
> 
> This ticket is about the latter, so we should do the sort first by compliance (same as in the first link) and then break ties by what is in the description.
> 
> Ask if you run into any other trouble!

Thanks,

I figured out this sorting function would work as in case of tie it will sort the tied items in descending order of their draft number

`Object.entries(implementation.results)
  .sort(
    (a, b) =>
      a[1].failedTests! +
        a[1].erroredTests! +
        a[1].skippedTests! -
        b[1].failedTests! -
        b[1].erroredTests! -
        b[1].skippedTests! ||
      +b[0].split(/draft(\d+-)?/).pop()! -
        +a[0].split(/draft(\d+-)?/).pop()!,
  )`

I used non null assertion because of Partial behaviour [here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/data/parseReportData.ts#L89-L91)

I want to make sure is this correct or I am missing something.

**Comment by sanskar-soni-9** [comment]
Thanks,

I figured out this sorting function would work as in case of tie it will sort the tied items in descending order of their draft number

```javascript
Object.entries(implementation.results)
  .sort(
    (a, b) =>
      a[1].failedTests! +
        a[1].erroredTests! +
        a[1].skippedTests! -
        b[1].failedTests! -
        b[1].erroredTests! -
        b[1].skippedTests! ||
      +b[0].split(/draft(\d+-)?/).pop()! -
        +a[0].split(/draft(\d+-)?/).pop()!, 
)
```

I used non null assertion because of Partial behaviour [here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/data/parseReportData.ts#L89-L91)

I want to make sure is this correct or I am missing something.

**Comment by Julian** [comment]
Thanks! That looks like the right idea -- but I think it'll be a bit more maintainable to avoid getting which dialects are newer from the name itself and rather explicitly add a field which is their release date.

Specifically, I suggest:

* Add a required `firstPublicationDate` field of type `Date` to [`Dialect` here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/data/Dialect.ts#L11)
* Set that field for all of the drafts in that file to when they were released. For ease, the release dates of all of the versions you'll find in that file are:
  - Draft 2020-12: January 28, 2020
  - Draft 2019-09: September 17, 2019
  - Draft 7: November 19, 2017
  - Draft 6: April 21, 2017
  - Draft 4: January 31, 2013
  - Draft 3: November 22, 2010
* Convert somehow between names (which is what you have in that Object) to `Dialect` objects -- either by passing the full dialect object in if that change is easy to make, or else simply by adding a lookup object -- in other words add a line to the `Dialect` file which is simply a mapping between dialect names like `draft2020-12` to `Dialect` objects, so that you can in your sort lookup the full `Dialect` object and get its `publicationDate`
* Use those dates to sort

Let me know if that makes sense but I think what you have is essentially functionally correct!

**Comment by sanskar-soni-9** [comment]
> Thanks! That looks like the right idea -- but I think it'll be a bit more maintainable to avoid getting which dialects are newer from the name itself and rather explicitly add a field which is their release date.
> 
> Specifically, I suggest:
> 
> * Add a required `firstPublicationDate` field of type `Date` to [`Dialect` here](https://github.com/bowtie-json-schema/bowtie/blob/a4cd53a8e44b2bcde4665fd73d14cd205403cf64/frontend/src/data/Dialect.ts#L11)
> * Set that field for all of the drafts in that file to when they were released. For ease, the release dates of all of the versions you'll find in that file are:
>   - Draft 2020-12: January 28, 2020
>   - Draft 2019-09: September 17, 2019
>   - Draft 7: November 19, 2017
>   - Draft 6: April 21, 2017
>   - Draft 4: January 31, 2013
>   - Draft 3: November 22, 2010
> * Convert somehow between names (which is what you have in that Object) to `Dialect` objects -- either by passing the full dialect object in if that change is easy to make, or else simply by adding a lookup object -- in other words add a line to the `Dialect` file which is simply a mapping between dialect names like `draft2020-12` to `Dialect` objects, so that you can in your sort lookup the full `Dialect` object and get its `firstPublicationDate`
> * Use those dates to sort, reversing the order
> 
> Let me know if that makes sense but I think what you have is essentially functionally correct!

Okay thanks alot, I was initially looking for the publication date in the data but couldn't find one so I used draft numbers. But this is of course the good way to solve the issue, will be raising a PR in an hour.

**Comment by Julian** [comment]
Fixed in #723.
